File: src/languages/gan/gan.ts
/**
 * Gan Chinese / 贛語 (gan), Nanchang 南昌 dialect — canonical IPA. The eighth Sinitic language (after Mandarin,
 * Cantonese, Wu, Min Nan, Jin, Hakka, and Xiang). Gan is a distinct primary branch of Sinitic spoken mainly in
 * Jiangxi; Nanchang (the provincial capital) is its prestige/representative variety — the Taiyuan-for-Jin /
 * Changsha-for-Xiang analogue.
 *
 * Written in Han characters. The reading dict (dict.tsv) already carries the Sinological IPA per syllable
 * (segmental IPA + a superscript pitch-number tone, narrow Nanchang diacritics kept verbatim), so the front-end is
 * the shared Han-dict engine (hanDictIpa.ts): greedy longest-match segmentation, superscript-tone → Chao contour
 * letters (SURFACE tone after a sandhi arrow ⁻), Han numerals. SINGLE authoritative source (Wiktionary/kaikki
 * Nanchang Sinological-IPA). ⚠ SINGLE-SOURCE: there is no independent referee, so nothing here is
 * cross-checked against a second transcription.
 */
import type { Phonemizer } from "../../registry.ts";
import { loadManifest } from "../../core/loadManifest.ts";
import { loadTsvMap } from "../../core/loadTsv.ts";
import { createHanDictPhonemizer, type ForeignPhonemizer, type HanDictDef, phonemizeHanWord } from "../sinitic/hanDictIpa.ts";

const DEF = loadManifest<HanDictDef>(import.meta.url, "gan.jsonc");

let DICT: Map<string, string> | undefined;
function dict(): Map<string, string> {
    return (DICT ??= loadTsvMap(import.meta.url, "dict.tsv"));
}

/** Build the Gan Chinese phonemizer. `foreign` handles embedded Latin runs. */
export function createGan(foreign?: ForeignPhonemizer): Phonemizer {
    return createHanDictPhonemizer(dict, DEF, foreign);
}

/** Bare word→IPA (tests / eval): a Han run → IPA. */
export function phonemizeWord(word: string): string {
    return phonemizeHanWord(dict, DEF, word);
}

File: csharp/Vernacula.Phonemizer/Languages/Gan/Gan.cs
/**
 * Gan Chinese / 贛語 (gan), Nanchang — canonical IPA. The dict already carries Sinological IPA per syllable, so
 * the front-end is the shared Han-dict engine (Core/HanDictIpa.cs); this file supplies the dict and manifest.
 * Ported from src/languages/gan/gan.ts — see that file for the sourcing.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Gan;

public static class GanPhonemizer
{
    private static readonly HanDictDef DEF = LoadManifest.Load<HanDictDef>("languages/gan", "gan.jsonc");

    private static IReadOnlyDictionary<string, string>? DICT;
    private static readonly object DICT_LOCK = new();
    private static IReadOnlyDictionary<string, string> Dict()
    {
        lock (DICT_LOCK) return DICT ??= LoadTsv.LoadTsvMap("languages/gan", "dict.tsv");
    }

    /** ⚠ THE NORMALIZER WRAPS the shared engine rather than being wired inside it — see the TS for why. */
    public static ILanguage CreateGan(Func<string, string>? foreign = null)
    {
        var engine = HanDictIpa.CreateHanDictPhonemizer(Dict, DEF, foreign);
        return new Wrapped(engine);
    }

    private sealed class Wrapped : ILanguage
    {
        private readonly ILanguage _engine;
        internal Wrapped(ILanguage engine) => _engine = engine;
        public string Text(string input) => _engine.Text(Normalize.NormalizeGan(input));
    }

    /** Bare word→IPA (tests / eval): a Han run → IPA. */
    public static string PhonemizeWord(string word) => HanDictIpa.PhonemizeHanWord(Dict, DEF, word);

    internal static void RegisterSelf() =>
        Registry.Register("gan", () => CreateGan(Registry.ReadAsEnglish));
}
